### PR TITLE
Feat/qwen35 tp phase1

### PR DESCRIPTION
**Pull request by Mrtroll486**
## Related

- Design doc PR: #450
- Related issues: #446, #459

## Background

This PR lands Qwen3.5-4B TP Phase 1.

Phase 1 is a correctness-first eager dense-TP milestone: it brings up TP2 dense full-attention / MLP sharding, worker/scheduler lifecycle, and the real serving path. This PR does not change the design contract doc `docs/models/qwen35/tp-design.md`; if review finds that additional contract wording is needed, I will update the design-doc branch / PR #450 rather than patching the design contract directly in this implementation PR.

Phase 2 design work, including TP unified steps, TP CUDA Graph, and sharded linear-attention/GDR state, should be decided in follow-up RFC issue discussion. Any unsettled design points mentioned below are follow-ups, not Phase 1 design commitments.

### Qwen3.5 TP runtime

- Added Qwen3.5 tensor-parallel config:
  - `rank/world_size`
  - dense-dimension divisibility checks
  - fail-closed startup for `TP > 1 && CUDA Graph`
  - unchanged TP1 default behavior
- Implemented dense TP weight loading:
  - sharded full-attention `q/k/v/o`
  - rank-local full-attention KV-head layout
  - sharded MLP `gate/up/down`
  - replicated embedding / tied `lm_head`
  - replicated linear-attention / GDR weights, conv state, and recurrent state for Phase 1
- Added a Qwen3.5 TP worker executor:
  - each rank worker owns its CUDA context, cuBLAS, NCCL comm, model shard, KV state, and recurrent/conv state
  - supports eager prefill, chunked prefill, and eager decode
  - `DropRequest` clears rank-local request state
- Routed the scheduler through a TP backend:
  - scheduler keeps logical request/page accounting
  - workers own physical rank-local KV/recurrent/conv state
  - TP scheduler currently selects eager prefill-only or eager decode-only steps
  - TP unified step is not enabled in Phase 1

### Qwen3.5 test framework

- Added shared TP2 test-device parsing through `OPENINFER_TEST_TP_DEVICES`.
  - Defaults to `0,1`.
  - Accepts explicit two-device pairs such as `1,2` or `2,3`.
  - Validates exactly two distinct CUDA ordinals.
  - Used by TP2 HF golden gates, TP2 scheduler e2e, and TP2 HTTP serving smoke.
- Extended Qwen3.5 test path configuration:
  - `OPENINFER_TEST_MODEL_PATH` selects the real Qwen3.5 weights path.
  - `OPENINFER_TEST_FRONTEND_MODEL_PATH` optionally selects frontend tokenizer/config metadata for HTTP serving tests.
  - `OPENINFER_TEST_FRONTEND_MODEL_PATH` falls back to `OPENINFER_TEST_MODEL_PATH` when unset.
- Refactored Qwen3.5 scheduler e2e so the same request-flow body can run against both:
  - TP1 single-GPU scheduler
  - TP2 public multi-device engine path
- Added TP2 HF golden-gate coverage:
  - short fixture replay through `Qwen35TpExecutor`
  - sequential and batched eager replay
  - long fixture replay through TP2 eager path
- Added TP scheduler/backend tests:
  - eager-only plan selection under TP
  - fail-closed `TP > 1 && CUDA Graph`
  - chunked prefill followed by eager decode
  - rank-local cleanup through request lifecycle paths
- Added a real TP2 HTTP serving integration test:
  - starts a Qwen3.5 TP2 engine
  - serves it through `openinfer_vllm_frontend::serve`
  - exercises OpenAI-compatible `/v1/models` and `/v1/completions`

### Server / frontend launch path

- Added Qwen3.5 launch options so the server path can pass:
  - `tp_size`
  - `device_ordinal`
  - `cuda_graph`
  - `max_prefill_tokens`
- Updated `openinfer-server` Qwen3.5 startup to use the new launch options.
  - `tp_size > 1` maps to devices `0..tp_size` in the server path.
  - CUDA Graph must be disabled for TP in Phase 1.
- Extended `bench_serving` with `--max-prefill-tokens` for Qwen3/Qwen3.5 chunked-prefill control.
- Added HTTP TP2 smoke coverage through the frontend path.
  - This PR does not directly change the `openinfer_vllm_frontend` crate.
  - It wires Qwen3.5 TP startup into the server/bench paths and validates frontend behavior end-to-end.

### Docs

- Added `docs/models/qwen35/tp-implementation.md`.
  - Records Phase 1 outcome.
  - Records validation evidence.
  - Records stable TP2 test knobs.
  - Records implementation constraints and follow-ups.
  - Does not replace the stable design contract in `tp-design.md`.

### TP2 correctness

- TP2 short HF logits gate passes:
  - sequential eager: `108` positions, mean `0.0258`, p99 `0.0801`, max `0.1298`
  - batched eager: `72` positions, mean `0.0257`, p99 `0.0809`, max `0.1298`
- TP2 long HF logits gate passes:
  - prompts `4097` / `8192`
  - sequential eager: `18` positions, mean `0.0232`, p99 `0.0792`, max `0.1035`

### Scheduler / serving

- TP2 scheduler e2e passes and covers:
  - context-window rejection
  - greedy / logprobs paths
  - sequential requests
  - repeated request reuse
  - concurrent mixed greedy/sampling requests
  - consumer drop
  - post-drop scheduler health
- TP2 HTTP serving smoke passes and covers:
  - `/v1/models`
  - non-streaming `/v1/completions`
  - streaming `/v1/completions`
  - concurrent completions
  - finite logprobs
  - chunked prefill forced with `max_prefill_tokens=1`
  - fail-closed startup for `TP2 + CUDA Graph`

### TP1 regression

- TP1 short/long HF logits gates pass
- TP1 scheduler e2e passes

### Compile / focused tests

- Qwen3.5 lib tests pass with real CUDA access.
- Qwen3.5 TP executor focused tests pass.
- Qwen3.5 scheduler focused tests pass.
- `serving_tp2` test target compiles.
- `openinfer-server --features qwen35-4b` test target compiles.

## Out Of Scope

- TP CUDA Graph capture/replay
- TP `RunUnifiedStep` mixed prefill+decode execution
- sharded linear-attention/GDR weights, kernels, conv state, or recurrent state
- vocab-parallel embedding / `lm_head`
- prefix cache / recurrent-state snapshots
- performance claims

### 1. TP unified step

The Phase 1 TP scheduler currently selects eager prefill-only or eager decode-only steps. It does not select a mixed prefill+decode unified step under TP.

This PR’s position is that Phase 1 should first prove dense TP correctness, worker-owned state, and rank-local cleanup/reuse. The semantics, scheduling policy, acceptance gates, and sequencing for TP `RunUnifiedStep` should be decided in a follow-up RFC issue.

If reviewers want `tp-design.md` to explicitly say that Phase 1 does not cover `RunUnifiedStep`, I will update the design-doc branch / PR #450 rather than changing the design contract directly in this PR.

### 2. Slot / slot-compaction wording

There are two concepts that are easy to conflate:
- single-GPU CUDA Graph decode slots: stable-address recurrent/conv slots in `BatchDecodeGraphState`
- PagedAttention KV pages/blocks: physical/logical KV cache pages

TP2 Phase 1 is eager-only and does not use TP CUDA Graph recurrent slots. It validates consistent rank-local request cleanup across finish/drop/cancel/reuse, which is not the same thing as TP CUDA Graph slot compaction.

If this distinction should become part of the design contract, I will update PR #450 / the design-doc branch.

### 3. Gated `q_proj` physical layout

The implementation confirmed that Qwen3.5 full-attention `q_proj` uses a head-interleaved physical layout: `[head q][head gate]`. TP sharding must preserve that layout. Rebuilding local shards as `[all local q][all local gate]` breaks numerics.

This is recorded in the implementation doc for now. If reviewers consider it part of the long-term partition contract, I will add the generic formula and TP2 example to PR #450 / the design-doc branch.

## Docs

Added:
- `docs/models/qwen35/tp-implementation.md`

This doc records the Phase 1 implementation and validation evidence. The stable design contract remains in `docs/models/qwen35/tp-design.md` / PR #450.

## Type of Change

- \[ ] Bug fix (non-breaking change which fixes an issue)

- \[x] New feature (non-breaking change which adds functionality)

- \[ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

- \[ ] Documentation update

## Checklist

- \[x] My code follows the style guidelines of this project (see `docs/conventions/coding-style.md`).

- \[x] I have performed a self-review of my own code.

- \[x] I have formatted my commits according to Commitizen conventions.

- \[x] I have run the local test suite and all tests pass (see `CLAUDE.md`).

### CODE REVIEW

File: openinfer-qwen35-4b/src/tp_executor.rs [hunk 1]
```diff
@@ -0,0 +1,1029 @@
...
+            let mut recurrent_refs = vec![&mut state.recurrent];
+            let logits = self.model.batch_prefill_logits(
+                &prompt,
+                std::slice::from_mut(&mut state.kv),
+                &mut recurrent_refs,
+            )?;
+
+            if chunk.finish_prefill {
+                if self.rank == 0 {
+                    let result = self.sample_final_prefill_chunk(chunk, &logits, sample_seed)?;
```
[the remark this answers is not in the document]
Comment (Mrtroll486) [reply openinfer-qwen35-4b/src/tp_executor.rs]: solved

File: openinfer-qwen35-4b/src/tp_executor.rs [hunk 2]
```diff
@@ -0,0 +1,1029 @@
...
+                    tensor_parallel: Some(TensorParallelConfig { rank, world_size }),
+                    device_ordinal,
+                },
+            )?);
+        }
+        let first = models
+            .first()
+            .ok_or_else(|| anyhow::anyhow!("Qwen3.5 TP executor loaded no models"))?;
+        let page_size = first.kv_pool().layout().page_size;
+        let capacity_pages_for_requests = first.kv_pool().capacity_pages().saturating_sub(1);
```
[the remark this answers is not in the document]
Comment (Mrtroll486) [reply openinfer-qwen35-4b/src/tp_executor.rs]: solved

**Comment by Mrtroll486** [comment]
@codex review

File: openinfer-kernels/build.rs [hunk 3]
```diff
@@ -1016,9 +1016,73 @@ fn generate_triton_artifacts(
...
+        .expect("generated Triton C source should declare one CUfunction");
+
+    let patched = src
+        .replace(
+            &format!("CUmodule {module_name} = NULL;"),
+            &format!("CUmodule {module_name}[16] = {{0}};"),
+        )
+        .replace(
+            &format!("CUfunction {function_name} = NULL;"),
+            &format!("CUfunction {function_name}[16] = {{0}};"),
```
[the remark this answers is not in the document]
Comment (Mrtroll486) [reply openinfer-kernels/build.rs]: This issue was not caused by switching from a single handle to a per-device list. The real fix was to support multiple CUDA devices with separate handles. We kept the fixed 16-entry table, and added a check that rejects any device ordinal greater than 15 before indexing it. If reviewers find this approach suboptimal, please let me know, and I will update it accordingly.

File: openinfer-kernels/build.rs [hunk 4]
```diff
@@ -1016,9 +1016,73 @@ fn generate_triton_artifacts(
...
+            &format!("cuFuncSetCacheConfig({function_name}, "),
+            &format!("cuFuncSetCacheConfig({function_name}[dev], "),
+        )
+        .replace(
+            &format!("cuFuncSetAttribute({function_name}, "),
+            &format!("cuFuncSetAttribute({function_name}[dev], "),
+        )
+        .replace(
+            &format!("if ({function_name} == NULL)\n       load_"),
+            &format!("int dev = 0;\n    CUDA_CHECK(cuCtxGetDevice(&dev));\n    if ({function_name}[dev] == NULL)\n       load_"),
```
[the remark this answers is not in the document]
Comment (Mrtroll486) [reply openinfer-kernels/build.rs]: solved

**Comment by CAICAIIs** [comment]
hi thanks your contribution, please fix the conflicts

**Comment by Mrtroll486** [comment]
@codex review

**Comment by Mrtroll486** [comment]
@codex review

File: openinfer-server/src/main.rs [hunk 5]
```diff
@@ -265,12 +265,16 @@ fn load_engine(args: &Args, model_type: ModelType) -> anyhow::Result<EngineHandl
...
-        ModelType::Qwen35 => openinfer_qwen35_4b::launch(
+        ModelType::Qwen35 => openinfer_qwen35_4b::launch_with_options(
             &args.model_path,
-            args.device_ordinal,
-            args.cuda_graph,
-            args.max_prefill_tokens
-                .unwrap_or(openinfer_qwen35_4b::DEFAULT_MAX_PREFILL_TOKENS),
+            openinfer_qwen35_4b::Qwen35LaunchOptions {
+                device_ordinal: args.device_ordinal,
+                tp_size: args.tp_size,
```
[the remark this answers is not in the document]
Comment (Mrtroll486) [reply openinfer-server/src/main.rs:310]: Solved.

File: openinfer-qwen35-4b/src/batch_decode.rs [hunk 6]
```diff
@@ -93,6 +93,9 @@ impl Qwen35Model {
         bufs: &mut BatchDecodeBuffers35,
     ) -> Result<()> {
         let eps = self.config.rms_norm_eps;
+        let tp = self.tensor_parallel;
+        let num_attention_heads = self.config.local_num_attention_heads(tp);
+        let num_key_value_heads = self.config.local_num_key_value_heads(tp);
```
[the remark this answers is not in the document]
Comment (Mrtroll486) [reply openinfer-qwen35-4b/src/batch_decode.rs:98]: Solved, this is introduced by rebase operation.

**Comment by Mrtroll486** [comment]
@codex review

File: openinfer-qwen35-4b/src/batch_decode.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Mrtroll486) [reply openinfer-qwen35-4b/src/batch_decode.rs:98]: solved

File: openinfer-qwen35-4b/src/scheduler.rs [hunk 7]
```diff
@@ -131,6 +153,544 @@ pub fn start_with_capacity(
...
+            scheduler_loop(
+                SchedulerBackend::Tp(backend),
+                submit_rx,
+                seed,
+                max_prefill_tokens,
+            );
+        })
+        .expect("failed to spawn Qwen3.5 TP scheduler thread");
+
+    Ok(SchedulerHandle::new_with_join_handle(submit_tx, join_handle).with_servable_len(servable))
```
[the remark this answers is not in the document]
Comment (Mrtroll486) [reply openinfer-qwen35-4b/src/scheduler.rs]: sloved.

**Comment by Mrtroll486** [comment]
All conflicts fixed.

File: openinfer-qwen35-4b/src/tp_executor.rs [hunk 8]
```diff
@@ -0,0 +1,1243 @@
...
+        }
+    }
+    Ok(())
+}
+
+fn wait_for_prefill(pending: Vec<mpsc::Receiver<Result<TpWorkerReply>>>) -> Result<PrefillResult> {
+    let mut result = None;
+    for (rank, recv) in pending.into_iter().enumerate() {
+        match recv
+            .recv()
```
Comment (CAICAIIs) [inline openinfer-qwen35-4b/src/tp_executor.rs]: A worker can return a rank-local CUDA/OOM error before its peers complete the corresponding NCCL collective. The controller then waits indefinitely on these unbounded recv() calls, and Drop can also block forever while joining a worker stuck in NCCL. Please add a runtime failure contract that aborts the communicator or process, marks the executor as poisoned, and prevents unbounded cleanup.

File: openinfer-qwen35-4b/src/tp_executor.rs [hunk 9]
```diff
@@ -0,0 +1,1243 @@
...
+
+    fn ensure_prefill_state(&mut self, request_id: RequestId) -> Result<usize> {
+        if let Some(idx) = self.request_index(request_id) {
+            return Ok(idx);
+        }
+        let state = TpRequestState {
+            request_id,
+            phase: TpRequestPhase::Prefilling,
+            kv: self.model.alloc_kv(),
+            recurrent: RecurrentState::new(self.model.device_ctx(), self.model.config())?,
```
Comment (CAICAIIs) [inline openinfer-qwen35-4b/src/tp_executor.rs:1141]: Each new TP request allocates a full replicated recurrent and convolution state on every rank. For Qwen3.5-4B this is about 49.125 MiB per request per rank, so the default `max_batch=64` requires about 3.07 GiB per rank after the KV pool has already consumed 85% of the available memory. Please reserve or preallocate this state before sizing the KV pool, or derive and advertise a lower capacity from the minimum available memory across ranks.

File: openinfer-qwen35-4b/src/tp_executor.rs (hunk 8 above)
Comment (Mrtroll486) [reply openinfer-qwen35-4b/src/tp_executor.rs]: Solved. The runtime fix adds a shared poison state, propagates worker errors and panics, uses bounded response and shutdown waits, and aborts the process if NCCL workers cannot exit, preventing indefinite runtime and cleanup hangs.

File: openinfer-qwen35-4b/src/tp_executor.rs (hunk 9 above)
Comment (Mrtroll486) [reply openinfer-qwen35-4b/src/tp_executor.rs:1141]: Solved. The capacity fix calculates per-request recurrent-state memory, reserves GPU runtime headroom, derives a safe max_batch for each rank, uses the minimum across all ranks, and rejects startup if even one request cannot fit.

File: openinfer-qwen35-4b/src/tp_executor.rs [hunk 10]
```diff
@@ -0,0 +1,1549 @@
...
+        requested_max_batch: usize,
+    ) -> Result<(Self, usize)> {
+        let cublas_guard = bind_worker_thread(&model)?;
+        let (free_bytes, total_bytes) = model
+            .device_ctx()
+            .ctx
+            .mem_get_info()
+            .map_err(|err| anyhow::anyhow!("failed to query TP rank {rank} memory: {err}"))?;
+        let recurrent_bytes = RecurrentState::allocation_bytes(model.config());
+        let max_batch = effective_recurrent_capacity(
```
Comment (CAICAIIs) [inline openinfer-qwen35-4b/src/tp_executor.rs:782]: This preserves only 512 MiB, but a legal 20k-token TP2 prefill needs about 2.66 GiB of scratch. Recurrent states can consume that headroom under high occupancy and cause OOM. Please subtract the configured prefill scratch before calculating capacity, or cap `max_prefill_tokens`.

Comment (Mrtroll486) [reply openinfer-qwen35-4b/src/tp_executor.rs:782]: Solved. Before calculating recurrent-state capacity, we subtract both the runtime reserve and the estimated prefill scratch allocation from the available memory. The scratch allocation is estimated using `min(max_prefill_tokens, PREFILL_CHUNK_LEN)`, where `PREFILL_CHUNK_LEN` is currently 20,000 tokens. Thus, we reserve scratch for the smaller of the configured maximum prefill budget and the hard cap of 20k tokens.

Note that the subtracted value is estimated value calculated by `GdrChunkwiseScratch35::estimate_bytes(...)`, this will be slightly bigger than 2.66 GiB.

**Review by CAICAIIs — APPROVED** [review]
